### PR TITLE
[SW-1648] Convert PySpark H2OFrame to DataFrame without Client

### PR DESCRIPTION
**Pull request by mn-mikke**
This PR depends on https://github.com/h2oai/h2o-3/pull/4090.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/h2o/utils/H2OContextRestAPIUtils.scala [hunk 1]
```diff
@@ -108,15 +149,14 @@ trait H2OContextRestAPIUtils extends H2OContextUtils {
   }
 
   private def getCloudInfoFromNode(endpoint: URI, conf: H2OConf): CloudV3 = {
-    val content = readStringURLContent(endpoint, "3/Cloud", conf)
-    new Gson().fromJson(content, classOf[CloudV3])
+    query[CloudV3](endpoint, "3/Cloud", conf)
   }
 
-  private def readStringURLContent(endpoint: URI, suffix: String, conf: H2OConf): String = {
+  private def query[ResultType : ClassTag](endpoint: URI, suffix: String, conf: H2OConf): ResultType = {
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/utils/H2OContextRestAPIUtils.scala]: 👍

File: core/src/main/scala/org/apache/spark/h2o/utils/ReflectionUtils.scala [hunk 2]
```diff
@@ -168,6 +170,14 @@ object ReflectionUtils {
     }
   }
 
+  def supportedType(columnType: String): SupportedType = columnType match {
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/utils/ReflectionUtils.scala]: I would maybe stress that this method converts H2O type to our supported type, what do you think?

File: core/src/main/scala/org/apache/spark/h2o/converters/H2OSparkEntity.scala [hunk 3]
```diff
@@ -77,5 +77,43 @@ private[converters] trait H2ORDDLike[T <: Frame] {
...
+  /** Chunk locations helps us to determine the node which really has the data we needs. */
+  override val chksLocation = if (isExternalBackend) Some(FrameUtils.getChunksLocations(frame)) else None
 }
+
+/**
+  * Contains functions that are shared between all REST-based H2ORDD types (i.e., Scala, Java)
+  */
+private[converters] trait H2ORESTBasedSparkEntity extends H2OSparkEntity {
+  /** Underlying H2O Frame */
+  @(transient @field @getter) val frame: H2OFrame
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2OSparkEntity.scala]: I think we can simplify to just `val frame: H2OFrame`

File: core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala [hunk 4]
```diff
@@ -219,7 +196,77 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@(transient @p
...
+  override val isExternalBackend = hc.getConf.runsInExternalClusterMode
+
+  // Get product type before building an RDD
+  def this(@transient frame: H2OFrame)
+          (@transient hc: H2OContext) = this(frame, ProductType.create[A])(hc)
+
+  protected override val colNames = frame.columns.map(_.name)
+
+  // Check that H2OFrame & given Scala type are compatible
+  if (!productType.isSingleton) {
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala]: NIT: There is some form of duplication with the similar code above

File: core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala [hunk 5]
```diff
@@ -219,7 +196,77 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@(transient @p
...
+/**
+  * Convert H2OFrame into an RDD (lazily).
+  *
+  * @param frame  an instance of H2O frame
+  * @param productType pre-calculated deconstructed type of result
+  * @param hc  an instance of H2O context
+  * @tparam A  type for resulting RDD
+  */
+private[spark]
+class H2ORESTRDD[A <: Product: TypeTag: ClassTag] private(@(transient @param @field) val frame: H2OFrame,
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala]: `@(transient @param @field) val frame` -> `val frame`

File: core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala [hunk 6]
```diff
@@ -71,39 +116,48 @@ class H2ODataFrame[T <: water.fvec.Frame](@transient val frame: T,
...
-    val iterator = new H2OChunkIterator[InternalRow] {
+/**
+  * H2O H2OFrame wrapper providing RDD[Row]=DataFrame API.
+  *
+  * @param frame frame which will be wrapped as DataFrame
+  * @param requiredColumns  list of the columns which should be provided by iterator, null means all
+  * @param hc an instance of H2O Context
+  */
+private[spark]
+class H2ORESTDataFrame(@transient val frame: H2OFrame, val requiredColumns: Array[String])
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala]: we can remove @transient

File: core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala [hunk 7]
```diff
@@ -71,39 +116,48 @@ class H2ODataFrame[T <: water.fvec.Frame](@transient val frame: T,
...
+                      (@transient val hc: H2OContext)
+  extends H2ODataFrameBase(hc.sparkContext) with H2ORESTBasedSparkEntity {
 
-      /** Frame reference */
-      override val keyName = frameKeyName
-      /** Processed partition index */
-      override val partIndex = split.index
+  def this(@transient frame: H2OFrame)
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala]: we can remove @transient

File: core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala [hunk 8]
```diff
@@ -71,39 +116,48 @@ class H2ODataFrame[T <: water.fvec.Frame](@transient val frame: T,
...
+    } else {
+      requiredColumns.map(colNames.indexOf)
+    }
+  }
 
-      override def next(): InternalRow = {
-        val row = readRow
-        converterCtx.increaseRowIdx()
-        row
+  override lazy val expectedTypes: Option[Array[Byte]] = {
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala]: Some duplication with code above

File: core/src/main/scala/ai/h2o/sparkling/frame/H2OFrame.scala [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+package ai.h2o.sparkling.frame
+
+import org.apache.spark.h2o.utils.NodeDesc
+
+/* This case class contains metadata describing H2O Frame accessed via REST API */
+case class H2OFrame(
+    frameId: String,
+    columns: Array[H2OColumn],
+    chunks: Array[H2OChunk]) {
+  @transient lazy val numberOfRows: Long = chunks.foldLeft(0L)((acc, chunk) => acc + chunk.numberOfRows)
```
[the remark this answers is not in the document]
Comment (mn-mikke) [reply core/src/main/scala/ai/h2o/sparkling/frame/H2OFrame.scala]: I use it as a pattern. I will remove it, thank you!

File: core/src/main/scala/org/apache/spark/h2o/converters/H2OSparkEntity.scala [hunk 10]
```diff
@@ -77,5 +77,43 @@ private[converters] trait H2ORDDLike[T <: Frame] {
...
+
+  /** Timestamp of the H2O Driver node */
+  override val driverTimeStamp: Short = -1 // Setting timestamp to -1 since there is no H2O client running
+
+  /** Cache frame key to get H2OFrame from the K/V store */
```
[the remark this answers is not in the document]
Comment (mn-mikke) [reply core/src/main/scala/org/apache/spark/h2o/converters/H2OSparkEntity.scala]: Good catch, thanks!

File: core/src/main/scala/ai/h2o/sparkling/frame/H2OFrame.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (jakubhava) [reply core/src/main/scala/ai/h2o/sparkling/frame/H2OFrame.scala]: Yup. The motivation for using @transient before was that H2OFrame client based had references to H2O & H2OContext. I think we avoid this with change :)

File: core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala [hunk 11]
```diff
@@ -257,16 +266,9 @@ class H2ORESTRDD[A <: Product: TypeTag: ClassTag] private(@(transient @param @fi
...
-  if (!productType.isSingleton) {
-    val productFields = productType.members.map(_.name)
-    val problems = productFields.diff(colNames).mkString(", ")
-
-    if (problems.nonEmpty) {
-      throw new IllegalArgumentException(s"The following fields are missing in frame: $problems; " +
-        "we have " + colNames.mkString(","))
-    }
+  protected override val colNames = {
+    val names =frame.columns.map(_.name)
```
Comment (jakubhava) [inline core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala]: NIT: missing space

**Pull request merged**